src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -530,49 +530,54 @@ private fun processItemDeclarations(
     originalProcessor: RsResolveProcessor,
     withPrivateImports: Boolean
 ): Boolean {
-    val macroItems = ItemsCache.build { add ->
-        for (macro in scope.macroCallList) {
-            macro.expansion?.forEach { add(it) }
-        }
+    val starImports = mutableListOf<RsUseItem>()
+    val itemImports = mutableListOf<RsUseItem>()
+
+    val directlyDeclaredNames = HashSet<String>()
+    val processor = { e: ScopeEntry ->
+        directlyDeclaredNames += e.name
+        originalProcessor(e)
     }
 
-    val (starImports, itemImports) = scope.useItemList.chain(macroItems.useItemList)
-        .filter { it.isPublic || withPrivateImports }
-        .partition { it.isStarImport }
+    fun processItem(item: RsItemElement): Boolean {
```

Comment:
The main trick is here: now we process items in source order, but we do postpone use item processing until the end.